rules_gapic/cpp/cc_gapic.bzl

```diff
@@ -0,0 +1,114 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("@com_google_api_codegen//rules_gapic:gapic.bzl", "gapic_srcjar", "unzipped_srcjar")
+
+def _cc_gapic_postprocessed_srcjar_impl(ctx):
+    gapic_zip = ctx.file.gapic_zip
+    output_main = ctx.outputs.main
+    output_main_h = ctx.outputs.main_h
+
+    output_dir_name = ctx.label.name
+    output_dir_path = "%s/%s" % (output_main.dirname, output_dir_name)
+
+    script = """
+    unzip -q {gapic_zip} -d {output_dir_path}
+    # TODO: Call formatter here
+    pushd .
+    cd {output_dir_path}
+    zip -q -r {output_dir_name}-h.srcjar . -i ./*.gapic.h
+    popd
+    mv {output_dir_path}/{output_dir_name}-h.srcjar {output_main_h}
+    cp {gapic_zip} {output_main}
+    """.format(
+        gapic_zip = gapic_zip.path,
+        output_dir_name = output_dir_name,
+        output_dir_path = output_dir_path,
+        output_main = output_main.path,
+        output_main_h = output_main_h.path,
+    )
+
+    ctx.actions.run_shell(
+        inputs = [gapic_zip],
+        command = script,
+        outputs = [output_main, output_main_h],
+    )
+
+_cc_gapic_postprocessed_srcjar = rule(
+    _cc_gapic_postprocessed_srcjar_impl,
+    attrs = {
+        "gapic_zip": attr.label(mandatory = True, allow_single_file = True),
+    },
+    outputs = {
+        "main": "%{name}.srcjar",
+        "main_h": "%{name}-h.srcjar",
+    },
+)
+
+def cc_gapic_srcjar(name, src, package, visibility = None):
+    raw_srcjar_name = "%s_raw" % name
+
+    gapic_srcjar(
+        name = raw_srcjar_name,
+        src = src,
+        package = package,
+        output_suffix = ".zip",
+        visibility = visibility,
+        gapic_generator = Label("//generator:protoc-gen-cpp_gapic"),
+    )
+
+    _cc_gapic_postprocessed_srcjar(
+        name = name,
+        gapic_zip = ":%s" % raw_srcjar_name,
+        visibility = visibility,
+    )
+
+def cc_gapic_library(name, src, package, deps = [], visibility = None):
```

Comment:
Caveat: This may not apply to Starlark. I don't know if its evaluation rules are the same as python's.

Making an parameter with a default value of the empty list is dangerous because the same list is reused across multiple function calls.
E.g.
```
def molluscs(squids=[]):
  squids.append("Humboldt")
  for s in squids:
    print(s)

mollusc()
mollusc()
```
The second call prints 
```
Humboldt
Humboldt
```

A pattern that I like is to use an empty tuple instead, which acts like an immutable list. It doesn't matter too much in this case because `deps` isn't mutated, but it's a pet peeve of mine.

Reply:
In Starlark `def cc_gapic_library(...)` technically is not even a function, it is a macro. The C++ templates is the closest analogy to it I can think of. Macros are expanded on loading phase (the first phase of bazel build process) like C++ templates generate actual C++ code on compile time. After that, what actually is being built by bazel is the expanded version of the macros (i.e. its internals).

Modifying input parameter is not a good practice in any language (I would guess), so it is not done here and I doubt will ever be done anywhere. If you do not have strong opinion about it I would keep this definition as is (there are many other rules  where it is used already (for other langs), and I would like to keep things consistent, and at long last it is Starlark, not Python).

Reply:
Good to know. Given how much Starlark looks like python, I was not expecting those semantics.